primitives/runtime-interface/src/lib.rs

```diff
@@ -153,6 +153,22 @@ pub use sp_std;
 ///         [17].to_vec()
 ///     }
 ///
+///     /// Call function, different version and only being registered.
+///     ///
+///     /// This `register_only` version is only being registered, aka exposed to the runtime,
+///     /// but the runtime will still use the version 2 of this function. This is useful for when
+///     /// new host functions should be introduced. Adding new host functions requires that all
+///     /// nodes have the host functions available, because otherwise they fail at instantiation
+///     /// of the runtime. With `register_only` the function will not be used when compiling the
+///     /// runtime, but it will already be there for a future version of the runtime that will
+///     /// switch to using these host function.
```

Comment:
Just to be sure I catch the intent and implication.
So the switch will be done by removing `register_only` and will require nodes updates (client hard fork).
So the PR makes it possible to stage some pending new function, but will still need a big update later.
The update may require some parachain to patch this flag until they do the client/node, can be a bit of work if some tests does need switching when removing the flag.
I feel like I am back in the situation where I would put this behind a feature (not a great idea I know).
Thinking twice, it feels good this way, the patch to retain `register` would be very small and the client update would in most case need to happen soon (to get latest runtime changes).

Reply:
The whole point of this is the following. Let's imagine we want to add new host functions, like with your trie pr. Currently we only we add these functions and directly start using them from the runtime. This way, the moment a runtime using these new host functions is enacted, the node needs to have this host functions ready or it starts to fail importing new blocks. With parachains and the relay chain node currently being compiled into the parachain node it even means that parachain teams need to bring out a new node release on time as well just to have the host functions ready for the relay chain. This is really shitty.

With `register_only`, we will be able to prepare the node side. Aka we will already be able to ship the new host functions now, but the runtime associated to the release or even the one after that don't use the new host functions. It will then be enabled at some later point when we are aware that most of the network has upgraded. 

This can not be done by features, because using a feature would completely remove the host function and again we would not have it in the node. The feature itself could then be used to disable the runtime functionality and maybe for also making removing the `register_only` when it is time.